constraint-solver/src/constraint_system.rs

```diff
@@ -0,0 +1,141 @@
+        )
+    }
+
+    pub fn iter_mut(&mut self) -> Box<dyn Iterator<Item = &mut V> + '_> {
+        Box::new(
+            [&mut self.bus_id, &mut self.multiplicity]
+                .into_iter()
+                .chain(self.payload.iter_mut()),
+        )
+    }
+}
+
+impl<T: FieldElement, V: Clone + Hash + Ord> BusInteraction<QuadraticSymbolicExpression<T, V>> {
+    /// Converts a bus interactions with fields represented by expressions
+    /// to a bus interaction with fields represented by range constraints.
+    pub fn to_range_constraints(
+        &self,
+        range_constraints: &impl RangeConstraintProvider<T, V>,
+    ) -> Option<BusInteraction<RangeConstraint<T>>> {
+        // TODO: Handle bus interactions with complex expressions.
+        Some(BusInteraction {
+            bus_id: expr_to_range_constraint(&self.bus_id, range_constraints)?,
+            payload: self
+                .payload
+                .iter()
+                .map(|expr| expr_to_range_constraint(expr, range_constraints))
+                .collect::<Option<Vec<_>>>()?,
+            multiplicity: expr_to_range_constraint(&self.multiplicity, range_constraints)?,
+        })
+    }
+}
+
+impl<T: FieldElement, V: Clone + Hash + Ord + Eq>
+    BusInteraction<QuadraticSymbolicExpression<T, V>>
+{
+    /// Attempts to execute the bus interaction, using the given
+    /// `BusInteractionHandler`. Returns a list of updates to be
+    /// executed by the caller.
+    pub fn solve(
+        &self,
+        bus_interaction_handler: &dyn BusInteractionHandler<T = T, V = V>,
+        range_constraints: &impl RangeConstraintProvider<T, V>,
+    ) -> Vec<Effect<T, V>> {
+        let Some(range_constraints) = self.to_range_constraints(range_constraints) else {
+            return vec![];
+        };
+        let range_constraints = bus_interaction_handler.handle_bus_interaction(range_constraints);
+        self.iter()
+            .zip(range_constraints.iter())
+            .filter_map(|(expr, rc)| {
+                if let Some(var) = expr.try_to_simple_unknown() {
+                    return Some(Effect::from_range_constraint_update(var, rc.clone()));
+                }
+                None
+            })
+            .collect()
+    }
+}
+
+/// A trait for handling bus interactions.
+pub trait BusInteractionHandler {
+    type T: FieldElement;
+    type V;
+
+    /// Handles a bus interaction, by transforming taking a bus interaction
+    /// (with the fields represented by range constraints) and returning
+    /// updated range constraints.
+    /// The range constraints are intersected with the previous ones by the
+    /// caller, so there is no need to do that in the implementation of this
+    /// trait.
+    fn handle_bus_interaction(
+        &self,
+        bus_interaction: BusInteraction<RangeConstraint<Self::T>>,
+    ) -> BusInteraction<RangeConstraint<Self::T>>;
+}
+
+fn expr_to_range_constraint<T: FieldElement, V: Clone + Hash + Ord + Eq>(
```

Comment:
I think I would make that a function on QuadraticSymbolicExpression that just multiplies the range constraints of all the components.

Reply:
Ok, I see that computing range constraints from complex expressions doesn't make a lot of sense for this application, but I don't see why we shouldn't have a function like that in QSE.

I mean something like

```
fn range_constraint(&self, range_constraints: impl ...) {
  let mut rc = constant.range_constraint();
  for (v, e) in self.linear {
    rc += range_constraints(v) * e.range_constraint();
  }
  // and also add the quadratic
  rc
}
``

Reply:
Makes sense! Let's do it when we need it :)